File: python/2179.count-good-triplets-in-an-array.py
class FenwickTree:
    def __init__(self, n):
        self.data = [0] * (n + 1)

    def update(self, i, delta):
        i += 1
        while i < len(self.data):
            self.data[i] += delta
            i += i & -i

    def query(self, i):
        i += 1
        res = 0
        while i > 0:
            res += self.data[i]
            i -= i & -i
        return res


class Solution:
    def goodTriplets(self, nums1: list[int], nums2: list[int]) -> int:
        count = 0
        n = len(nums1)

        pos2 = {v: i for i, v in enumerate(nums2)}
        reverseIndexMapping = {pos2[v]: i for i, v in enumerate(nums1)}

        ft = FenwickTree(n)

        for idx in range(n):
            pos = reverseIndexMapping[idx]
            left = ft.query(pos)
            ft.update(pos, 1)
            right = n - 1 - pos - (idx - left)

            count += left * right

        return count

File: cpp/2179.count-good-triplets-in-an-array.cpp
#include <vector>

using std::vector;

class FenWick {
private:
  vector<int> data;

public:
  FenWick(int n) : data(n + 1, 0) {}

  void update(int idx, int val) {
    idx++;
    while (idx < data.size()) {
      data[idx] += val;
      idx += idx & -idx;
    }
  }

  int query(int idx) {
    idx++;
    int res = 0;
    while (idx > 0) {
      res += data[idx];
      idx -= idx & -idx;
    }
    return res;
  }
};

class Solution {
public:
  long long goodTriplets(vector<int> &nums1, vector<int> &nums2) {
    long count = 0;
    const int n = nums1.size();

    vector<int> pos2(n), reverseIndexMapping(n);
    for (int i = 0; i < n; i++) {
      pos2[nums2[i]] = i;
    }
    for (int i = 0; i < n; i++) {
      reverseIndexMapping[pos2[nums1[i]]] = i;
    }

    FenWick fenwick(n);
    for (int i = 0; i < n; i++) {
      int pos = reverseIndexMapping[i];
      int left = fenwick.query(pos);
      fenwick.update(pos, 1);

      int right = n - 1 - pos - i + left;

      count += long(left) * right;
    }

    return count;
  }
};